pkg/store/proxy.go

```diff
@@ -274,6 +277,80 @@ func (s *ProxyStore) TSDBInfos() []infopb.TSDBInfo {
 	return infos
 }
 
+func (s *ProxyStore) LabelNamesBloom() bloom.Filter {
+	s.mtx.Lock()
+	defer s.mtx.Unlock()
+
+	return s.labelNamesBloom
+}
+
+func (s *ProxyStore) UpdateLabelNamesBloom(ctx context.Context) error {
+	var (
+		names          []string
+		mtx            sync.Mutex
+		g, gctx        = errgroup.WithContext(ctx)
+		storeDebugMsgs []string
+	)
+
+	for _, st := range s.stores() {
+		st := st
+
+		if s.debugLogging {
+			storeDebugMsgs = append(storeDebugMsgs, fmt.Sprintf("Store %s queried", st))
+		}
+
+		g.Go(func() error {
+			resp, err := st.LabelNames(gctx, &storepb.LabelNamesRequest{})
+			if err != nil {
+				return errors.Wrapf(err, "fetch label names from store for updating bloom filter %s", st)
+			}
+
+			mtx.Lock()
+			if len(names) != 0 {
+				names = union(names, resp.Names)
+			} else {
+				names = resp.Names
+			}
+			mtx.Unlock()
+
+			return nil
+		})
+	}
+
+	if err := g.Wait(); err != nil {
+		return err
+	}
+
+	level.Debug(s.logger).Log("msg", strings.Join(storeDebugMsgs, ";"))
+
+	s.mtx.Lock()
+	s.labelNamesBloom = bloom.NewFilterForStrings(names...)
+	s.mtx.Unlock()
+
+	return nil
+}
+
+func union(sliceA, sliceB []string) []string {
```

Comment:
I think this is quite inefficient because we constantly convert between a map and a slice. Can we maintain a single map and add keys to it, and then at the end call `maps.Keys` on it?